src/frontend/src/optimizer/plan_node/logical_recursive_union.rs

```diff
@@ -0,0 +1,124 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use pretty_xmlish::{Pretty, XmlNode};
+use risingwave_common::bail_not_implemented;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+use smallvec::{smallvec, SmallVec};
+
+use super::expr_visitable::ExprVisitable;
+use super::generic::GenericPlanRef;
+use super::utils::{childless_record, Distill};
+use super::{
+    generic, ColPrunable, ColumnPruningContext, ExprRewritable, Logical, PlanBase, PlanTreeNode,
+    PredicatePushdown, PredicatePushdownContext, RewriteStreamContext, ToBatch, ToStream,
+    ToStreamContext,
+};
+use crate::binder::ShareId;
+use crate::error::Result;
+use crate::utils::Condition;
+use crate::PlanRef;
+
+/// `LogicalRecursiveUnion` returns the union of the rows of its inputs.
+/// note: if `all` is false, it needs to eliminate duplicates.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalRecursiveUnion {
+    pub base: PlanBase<Logical>,
+    core: generic::RecursiveUnion<PlanRef>,
+}
+
+impl LogicalRecursiveUnion {
+    pub fn new(base_plan: PlanRef, recursive: PlanRef, id: ShareId) -> Self {
+        let core = generic::RecursiveUnion {
+            base: base_plan,
+            recursive,
+            id,
+        };
+        let base = PlanBase::new_logical_with_core(&core);
+        LogicalRecursiveUnion { base, core }
+    }
+
+    pub fn create(base_plan: PlanRef, recursive: PlanRef, id: ShareId) -> PlanRef {
+        Self::new(base_plan, recursive, id).into()
+    }
+
+    pub(super) fn pretty_fields(base: impl GenericPlanRef, name: &str) -> XmlNode<'_> {
+        childless_record(name, vec![("id", Pretty::debug(&base.id().0))])
+    }
+}
+
+impl PlanTreeNode for LogicalRecursiveUnion {
+    fn inputs(&self) -> SmallVec<[PlanRef; 2]> {
+        smallvec![self.core.base.clone(), self.core.recursive.clone()]
+    }
+
+    fn clone_with_inputs(&self, inputs: &[PlanRef]) -> PlanRef {
+        let mut inputs = inputs.iter().cloned();
+        Self::create(inputs.next().unwrap(), inputs.next().unwrap(), self.core.id)
+    }
+}
+
+impl Distill for LogicalRecursiveUnion {
+    fn distill<'a>(&self) -> XmlNode<'a> {
+        Self::pretty_fields(&self.base, "LogicalRecursiveUnion")
+    }
+}
+
+impl ColPrunable for LogicalRecursiveUnion {
+    fn prune_col(&self, required_cols: &[usize], ctx: &mut ColumnPruningContext) -> PlanRef {
+        let new_inputs = self
+            .inputs()
+            .iter()
+            .map(|input| input.prune_col(required_cols, ctx))
+            .collect_vec();
+        let new_plan = self.clone_with_inputs(&new_inputs);
+        self.ctx()
+            .insert_rcte_cache_plan(self.core.id, new_plan.clone());
+        new_plan
+    }
```

Comment:
Here we reuse the `ShareId` but insert a `new_plan` which might have a different schema from the `old_plan`, but `cte_ref` has a `base` field which represents a plan at the beginning of planning, so it would be inconsistent with the `rcte_cache_plan` updated here. That's why I think we should never use `base` of `cte_ref` after the initial planing.

Reply:
> but cte_ref has a base field which represents a plan at the beginning of planning, so it would be inconsistent with the rcte_cache_plan updated here.

agree. the only essential use case for `base` is to get the current `OptimizerContext`, for `GenericPlanNode`'s impl methods I think we could just stick with the current `stream_key`'s logic - in this case everything is consistent with the on-the-fly plan cache stored in the context.

Reply:
implementation of the current `stream_key` is here (the consistent one we should stick to): https://github.com/risingwavelabs/risingwave/pull/16680/files#diff-1d6ad1cdae116ec20167aafcc080d20606c2fb3da492cea9c6a398412456d96bR63.